packages/data-context/src/sources/RelevantRunSpecsDataSource.ts

```diff
@@ -136,106 +80,198 @@ export class RelevantRunSpecsDataSource {
-      const runSpecsToReturn: RunSpecReturn = {
-        runSpecs: {},
-        statuses: {},
-        testCounts: {},
-      }
+  pollForSpecs (runId: string, info: GraphQLResolveInfo) {
+    debug(`pollForSpecs called`)
+    //TODO Get spec counts before poll starts
+    if (!this.#poller) {
+      this.#poller = new Poller(this.ctx, 'relevantRunSpecChange', this.#pollingInterval, async (subscriptions) => {
+        debug('subscriptions', subscriptions)
+        const runIds = uniq(compact(subscriptions?.map((sub) => sub.meta?.runId)))
 
-      const { current, next } = cloudProject
+        debug('Polling for specs for runs: %o - runIds: %o', runIds)
 
-      const formatCloudRunInfo = (cloudRunDetails: Partial<CloudRun>) => {
-        const { runNumber, totalInstanceCount, completedInstanceCount } = cloudRunDetails
+        const query = this.createQuery(compact(subscriptions.map((sub) => sub.meta?.info)))
 
-        if (runNumber && isValidNumber(totalInstanceCount) && isValidNumber(completedInstanceCount)) {
-          return {
-            totalSpecs: totalInstanceCount,
-            completedSpecs: completedInstanceCount,
-            runNumber,
-            scheduledToCompleteAt: cloudRunDetails.scheduledToCompleteAt,
-          }
-        }
+        //debug('query', query)
 
-        return undefined
-      }
+        this.#query = query
 
-      if (current && current.status && current.totalTests !== null) {
-        runSpecsToReturn.runSpecs.current = formatCloudRunInfo(current)
-        runSpecsToReturn.statuses.current = current.status
-        runSpecsToReturn.testCounts.current = current.totalTests
-      }
+        const runs = await this.getRelevantRunSpecs(runIds)
 
-      if (next && next.status && next.totalTests !== null) {
-        runSpecsToReturn.runSpecs.next = formatCloudRunInfo(next)
-        runSpecsToReturn.statuses.next = next.status
-        runSpecsToReturn.testCounts.next = next.totalTests
-      }
+        debug(`Run data is `, runs)
+
+        runs.forEach(async (run) => {
+          const cachedRun = this.#cached.get(run.id)
+
+          if (!cachedRun || !isEqual(run, cachedRun)) {
+            debug(`Caching for id %s: %o`, run.id, run)
+            this.#cached.set(run.id, { ...run })
+
+            const cachedRelevantRuns = this.ctx.relevantRuns.cache
 
-      return runSpecsToReturn
+            if (run.runNumber === cachedRelevantRuns.selectedRunNumber) {
+              const projectSlug = await this.ctx.project.projectId()
+
+              debug(`Invalidate cloudProjectBySlug ${projectSlug}`)
+
+              await this.ctx.cloud.invalidate('Query', 'cloudProjectBySlug', { slug: projectSlug })
+              await this.ctx.emitter.relevantRunChange(cachedRelevantRuns)
+            }
+
+            this.ctx.emitter.relevantRunSpecChange(run)
+          }
+        })
+      })
     }
 
-    return SPECS_EMPTY_RETURN
+    const filter = (run: PartialCloudRunWithId) => {
```

Comment:
I have added some JSDocs to the `DataEmitterActions.subscribeTo` method to better describe the options. This filter is a predicate to allow for filtering the values being "emitted" for a subscription.  Since there can be multiple "subscriptions" tied to the same "event", I needed a way to only send certain values back on certain subscriptions.  

In the Run Navigation use case, there could be multiple running builds in the Cloud at one time that the component wants to subscribe to.  The backend query will combine them together and execute one query to the Cloud using the `cloudNodesById` field.  When the result comes back, the code iterates over each `CloudRun` returned to check the cache to see if it changes.  If it has changed and an event is sent to the emitter, the `filter` will only send through the runs with the matching `id` to the correct upstream (i.e. front end) subscription.